### ISSUE TITLE
Right clicking auto choose only choice (75-80%)

### ISSUE THREAD

**Issue by skaviouz** [Sigil-Ebook/Sigil#499]
1.1.0 / 5.12.6 / 2020.02.07
windows 10 - laptop @ 1366x768 res

suggestion: move the right click menu 1 or 2 pixels to the right if there is room.

see #454 as it has a similar title / issue but is different. - edit - i renamed my issue.

If you have two+ tabs opened and you right click one of the tabs one of two things may happen.

1. all other tabs close
2. right click menu dialogue opens with an item 'Close Other Tabs'

I've tested 30 times and its ~7:23~ 23:7 so It's probably a 20-25% chance for the menu to be opened for me.

If I do not right click, but hold the right mouse button it appears that 20/25% of the time the menu item will not be highlighted with the dialogue menu opening as it was intended. If you move the mouse cursor out of bounds of that context menu after that menu item was highlighted (intentionally hovering over it when it was not selected initially too) then the right click menu closes when the right click button is released.

edit: also if you move the cursor away before the menu finishes rendering (a few ms) then I can get a 99% chance of the menu staying open as the first item is never hovered over.

it should be consistent, so either have it always close or not.

I was working and when the tabs closed on me i did freak out, but im not angry because it did auto save. so I thought that I would make a suggestion considering that I have been using sigil for a long time. Thanks :)

TLDR; suggestion: move the right click menu 1 or 2 pixels to the right if there is room, or heck even 1 or 2 pixels to the left to make it so it always has the first item selected so we get that special "consistency".

**Comment by dougmassay** [comment]
Turn off HighDPI scaling in the UI preferences and see if that helps. May need to resize some widgets afterwards, but I've found that on smaller laptop screens, Qt's automatic scaling doesn't always work so well with all resolutions.

**Comment by dougmassay** [comment]
That's in Sigil's preferences, just to be clear. I'm not suggesting any OS-level changes.

**Comment by BeckyDTP** [comment]
I always have HighDPI turned off.

The problem with tabs also annoyed me, so I added a piece of code to TabBar.cpp file:
`    QAction *saveMyTabs = new QAction(tr("Save My Tabs!"), menu);`
`    menu->addAction(saveMyTabs);`
`    menu->addSeparator();`

![save-my-tabs](https://user-images.githubusercontent.com/22416021/75242122-c432e680-57c7-11ea-9fd2-cd99934b5613.png)

This is obviously just a workaround, but it works.

Unfortunately, I can't move this submenu to the right by 2px, although it may be possible.

**Comment by dougmassay** [comment]
I'm not quite sure how moving the context menu (which wouldn't likely be trivial at all to do) would help if the context menu never appears when the problem manifests. This is clearly a Qt bug related to existing bugs involving clicking on inactive tabs in my opinion. Saw the exact same kind of user report in calibre's support forums when right-clicking tabs.

@BeckyDTP , what is it that adding the empty menu item/action achieves here? I'm confused. Does adding it make the context menu appear more regularly, or does it just make it harder to accidentally select "Close Other Tabs?"

 Does it manifest if you activate the tab first (by left-clicking on it) before right-clicking the tab in question? I freely admit that I don't really find that context menu item very handy at all. So I never right-click on tabs in the first place. Especially given Qt's increasing tab flakiness in newer versions.

I'll test if disabling the ability to reorder tabs helps with this. If so, users can make a choice between rearrangeable tabs with flaky behavior, or non-rearrangeable tabs with stable behavior.

**Comment by BeckyDTP** [comment]
> @BeckyDTP , what is it that adding the empty menu item/action achieves here? I'm confused. Does adding it make the context menu appear more regularly, or does it just make it harder to accidentally select "Close Other Tabs?"

Yes, just make it to accidentally select "Close Other Tabs".

The problem exists for both the active and all inactive tabs.

I just noticed that there is a similar problem in the Caliber editor, although there are two menu items, so only one tab may accidentally close.
* Close Tab
* Close All Tabs

**Comment by dougmassay** [comment]
Yes. That calibre editor issue is the one I was referring to.

**Comment by dougmassay** [comment]
And now for the $64,000 questions (because you really didn't think I WASN'T going to ask them did you?). ;)

Once you know that there's only one possible action that can be accomplished by right-clicking a Code View tab, why would you 1) right-click a tab when you didn't want all of the other tabs closed, or 2) mind when the action you WANTED to happen, happened one-step sooner?

It's got nothing to do with the position of the context tab, and has everything to do with the mouse not moving when right-clicking. And for the record, I have yet to be able to make the active tab "accidentally" close when right-clicking on tabs in Windows. I can only replicate the issue when when-right-clicking on _inactive_ tabs (and only then when slightly twitching the mouse while right-clicking). Flaky inactive tabs have been an issue with Qt for a long time. You just have to slow down and not move the mouse at all while clicking (left or right).

**Comment by dougmassay** [comment]
Also: clicking closer the top of the tab should help workaround the issue.

**Comment by dougmassay** [comment]
Definitely way worse on Windows than other platforms

**Comment by dougmassay** [comment]
I see some options to try. Let me do a little experimenting tonight.

**Comment by dougmassay** [comment]
@BeckyDTP can you give this commit of mine a try and see if it eliminates the problem for you?

2ad82a4dcd184da07736719b797bebbac36cbdb8

**Comment by BeckyDTP** [comment]
Done.
For me (Windows 10 Pro 1909) – super!
Thank you.

**Comment by BeckyDTP** [comment]
Unfortunately, when you increase the number of attempts, there is still a problem with autoclick and all tabs close. It is difficult for me to provide the exact percentage of "bad" clicks, but this shift probably does not work correctly. I changed the value of the downward shift to 10px and the menu did not appear below, but still in the same place.

Interesting. I will do some more experiments.

**Comment by dougmassay** [comment]
Are you certain? I've clicked hundreds of tabs hundreds of times after this change and I can't get the problem to happen anymore. At all. And it was happening A LOT for me on this particular machine. Nearly all the time in fact. I can't even get it to accidentally highlight the action without working extremely hard to do so.

**Comment by BeckyDTP** [comment]
I'm sure.
I have the current build with your latest commit.
The first attempt was successful and I was happy.
Then I did the second, third, fourth, fifth attempt - but it doesn't work (to be precise: it doesn't always work).

**Comment by dougmassay** [comment]
I got nothing. There's nothing in there that should be subject to a random or cumulative affect. The menu is simply reconstructed each and every time a tabbar is right-clicked. As I mentioned, I've gone absolutely and ridiculously right-click-happy since the change, and I've not had it happen yet.

**Comment by dougmassay** [comment]
And for the record... this fix is never going to completely eliminate the possibility of someone holding the right mouse button down too long, and moving the mouse just enough to highlight the menu item. It's only meant to greatly reduce the likelihood of it accidentally happening almost every single time. Once the menu item gets highlighted with the mouse button still down, there's no stopping the event from firing whenever the mouse button does get released.

**Comment by BeckyDTP** [comment]
I have reduced the sensitivity settings in the mouse driver and now it's good! Maybe it is a problem with my hand that not all attempts were successful.

So I confirm again that the change in your commit works great.

**Comment by dougmassay** [comment]
Thanks. I'll still keep this open for a while in case something else comes to light.

**Comment by dougmassay** [comment]
@skaviouz : there's a ci build of this commit if you want to see if my fix works for you. Keep in mind that it's beta and will have a few incomplete features we're currently working on. Other than that, it should still be a working version of Sigil. Your call.

https://drive.google.com/uc?id=1L1ayH4aUAS_im-FdQ4qC1vTGzVqh9JSm&export=download

**Comment by kevinhendricks** [comment]
FWIW, macOS exhibits none of these issues on active or inactive tabs.  So the bug in Qt probably happens in the platform specific code or Style someplace and not in the general QTabBar related code.

**Comment by dougmassay** [comment]
Calling this one done. The fix will be in the next release. Thanks for the report!
